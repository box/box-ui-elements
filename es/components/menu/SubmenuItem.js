const _excluded = ["children", "className", "isDisabled"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import * as React from 'react';
import classNames from 'classnames';
import omit from 'lodash/omit';
import debounce from 'lodash/debounce';
import './SubmenuItem.scss';
import Arrow16 from '../../icon/fill/Arrow16';
const SUBMENU_LEFT_ALIGNED_CLASS = 'is-left-aligned';
const SUBMENU_BOTTOM_ALIGNED_CLASS = 'is-bottom-aligned';
const SUBMENU_RIGHT_BOTTOM_ALIGNED_CLASS = 'is-right-bottom-aligned';
/**
 * A menu-item component which triggers open a submenu
 *
 * @NOTE: Nested submenus are NOT currently supported, switching
 * focus with arrow keys in the subsubmenu is not working properly.
 */
class SubmenuItem extends React.Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "state", {
      isSubmenuOpen: false,
      submenuFocusIndex: null
    });
    _defineProperty(this, "getMenuAlignmentClasses", () => {
      if (!this.submenuTriggerEl || !this.submenuEl) {
        return {};
      }
      const {
        rightBoundaryElement,
        bottomBoundaryElement
      } = this.props;
      const submenuElBounding = this.submenuEl.getBoundingClientRect();
      const submenuTriggerElBounding = this.submenuTriggerEl.getBoundingClientRect();
      const rightBoundaryElementBounding = rightBoundaryElement ? rightBoundaryElement.getBoundingClientRect() : {
        right: window.innerWidth
      };
      const bottomBoundaryElementBounding = bottomBoundaryElement ? bottomBoundaryElement.getBoundingClientRect() : {
        bottom: window.innerHeight
      };
      const isLeftAligned = submenuTriggerElBounding.right + submenuElBounding.width > rightBoundaryElementBounding.right;
      const isBottomAligned = submenuTriggerElBounding.top + submenuElBounding.height > bottomBoundaryElementBounding.bottom;
      const isRightBottomAligned = submenuTriggerElBounding.bottom + submenuElBounding.height > bottomBoundaryElementBounding.bottom;
      return {
        [SUBMENU_LEFT_ALIGNED_CLASS]: isLeftAligned,
        [SUBMENU_BOTTOM_ALIGNED_CLASS]: isBottomAligned,
        [SUBMENU_RIGHT_BOTTOM_ALIGNED_CLASS]: isRightBottomAligned // Used only in medium-screen viewport sizes
      };
    });
    _defineProperty(this, "handleMenuItemClick", event => {
      const {
        isDisabled,
        onClick
      } = this.props;

      // If aria-disabled is passed as a prop, we should ignore clicks on this menu item
      if (isDisabled) {
        event.stopPropagation();
        event.preventDefault();
        return;
      }
      if (onClick) {
        onClick(event);
      }

      // If event target is triggering submenu element, do not propagate to close menu
      if (this.submenuEl && !this.submenuEl.contains(event.target)) {
        event.stopPropagation();
        event.preventDefault();
      }
    });
    _defineProperty(this, "handleKeyDown", event => {
      switch (event.key) {
        case ' ':
        case 'Enter':
        case 'ArrowRight':
          event.stopPropagation();
          event.preventDefault();
          this.openSubmenuAndFocus();
          break;
        default:
          break;
      }
    });
    _defineProperty(this, "closeSubmenu", debounce(() => {
      this.setState({
        isSubmenuOpen: false
      });
    }, 50));
    _defineProperty(this, "closeSubmenuAndFocusTrigger", isKeyboardEvent => {
      this.closeSubmenu();
      if (this.submenuTriggerEl && isKeyboardEvent) {
        this.submenuTriggerEl.focus();
      }
    });
    _defineProperty(this, "openSubmenu", () => {
      this.closeSubmenu.cancel();
      const {
        onOpen
      } = this.props;
      if (onOpen) {
        onOpen();
      }
      this.setState({
        isSubmenuOpen: true,
        submenuFocusIndex: null
      });
    });
    _defineProperty(this, "openSubmenuAndFocus", () => {
      const {
        onOpen
      } = this.props;
      if (onOpen) {
        onOpen();
      }
      this.setState({
        isSubmenuOpen: true,
        submenuFocusIndex: 0
      });
    });
  }
  render() {
    const _this$props = this.props,
      {
        children,
        className,
        isDisabled
      } = _this$props,
      rest = _objectWithoutProperties(_this$props, _excluded);
    const {
      isSubmenuOpen,
      submenuFocusIndex
    } = this.state;
    const elements = React.Children.toArray(children);
    const submenuTriggerContent = elements[0];
    const submenu = elements[1];
    if (elements.length !== 2 || !submenuTriggerContent || !submenu) {
      throw new Error('SubmenuItem must have exactly two children, a trigger component and a <Menu>');
    }
    const chevron = /*#__PURE__*/React.createElement(Arrow16, {
      className: "menu-item-arrow",
      width: 12,
      height: 12
    });
    const menuItemProps = _objectSpread(_objectSpread({}, omit(rest, ['bottomBoundaryElement', 'onClick', 'onOpen', 'rightBoundaryElement', 'role', 'tabIndex'])), {}, {
      'aria-disabled': isDisabled ? 'true' : undefined,
      'aria-expanded': isSubmenuOpen ? 'true' : 'false',
      'aria-haspopup': 'true',
      className: classNames('menu-item', 'submenu-target', className),
      onClick: this.handleMenuItemClick,
      onMouseLeave: this.closeSubmenu,
      onMouseEnter: this.openSubmenu,
      onKeyDown: this.handleKeyDown,
      ref: ref => {
        this.submenuTriggerEl = ref;
      },
      role: 'menuitem',
      tabIndex: -1
    });
    const submenuProps = {
      className: classNames(submenu.props.className, 'submenu', this.getMenuAlignmentClasses()),
      initialFocusIndex: submenuFocusIndex,
      // Hide the menu instead of unmounting it. Otherwise onMouseLeave won't work.
      isHidden: !isSubmenuOpen,
      isSubmenu: true,
      onClose: this.closeSubmenuAndFocusTrigger,
      setRef: ref => {
        this.submenuEl = ref;
      }
    };
    return /*#__PURE__*/React.createElement("li", menuItemProps, submenuTriggerContent, chevron, /*#__PURE__*/React.cloneElement(submenu, submenuProps));
  }
}
export default SubmenuItem;
//# sourceMappingURL=SubmenuItem.js.map