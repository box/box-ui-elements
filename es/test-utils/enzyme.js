function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import { mount as baseMount, shallow as baseShallow } from 'enzyme';

// Global providers that need setup
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../styles/theme';

// Wrap this around every component so they can use app-wide context.
// Prevents breaking due to missing providers.
// See https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/getWrappingComponent.md
/* eslint-disable react/prop-types */
const Wrappers = ({
  children
}) => {
  return /*#__PURE__*/React.createElement(ThemeProvider, {
    theme: defaultTheme
  }, children);
};

/**
 * mount() from Enzyme but with required app-wide context providers included by default
 */
const mountConnected = (element, options = {}) => baseMount(element, _objectSpread({
  wrappingComponent: Wrappers
}, options));

/**
 * shallow() from Enzyme but with required app-wide context providers included by default.
 */
const shallowConnected = (element, options = {}) => baseShallow(element, _objectSpread({
  wrappingComponent: Wrappers
}, options));

// Re-export everything from Enzyme
export * from 'enzyme';
// Export wrapped methods
export { mountConnected, shallowConnected };
//# sourceMappingURL=enzyme.js.map