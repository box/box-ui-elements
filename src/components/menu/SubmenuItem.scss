@import '../../styles/variables';

.submenu-target {
    $bdl-submenu-offset: $bdl-grid-unit * -3;

    position: relative;

    .submenu {
        position: absolute;
        top: $bdl-submenu-offset;
        right: auto;
        left: calc(100% - #{$bdl-submenu-offset});

        &.is-left-aligned {
            right: 100%;
            left: auto;
        }

        &.is-hidden {
            right: auto;
            left: -9999px;
        }

        &.is-bottom-aligned {
            top: auto;
            bottom: $bdl-submenu-offset;
        }
    }

    .menu-item-arrow {
        position: absolute;
        top: $bdl-grid-unit * 3;
        right: $bdl-grid-unit * 2;
    }
}
