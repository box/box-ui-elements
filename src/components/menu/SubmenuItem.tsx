import * as React from 'react';
import classNames from 'classnames';
import omit from 'lodash/omit';
import debounce from 'lodash/debounce';

import './SubmenuItem.scss';
import Arrow16 from '../../icon/fill/Arrow16';

const SUBMENU_LEFT_ALIGNED_CLASS = 'is-left-aligned';
const SUBMENU_BOTTOM_ALIGNED_CLASS = 'is-bottom-aligned';

export interface SubmenuItemProps {
    /** 'aria-disabled' - ARIA attribute describing whether the submenu item is disabled */
    'aria-disabled'?: boolean | 'true' | 'false';
    /** 'aria-disabled' - ARIA attribute describing whether the submenu item is expanded */
    'aria-expanded'?: boolean | 'true' | 'false';
    /** 'aria-disabled' - ARIA attribute describing whether the submenu item has a popup */
    'aria-haspopup'?: boolean | 'true' | 'false';
    /** bottomBoundaryElement - an HTMLElement defining the bottom boundary for the submenu item */
    bottomBoundaryElement?: HTMLElement;
    /** children - submenu item content */
    children?: Array<React.ReactChild> | React.ReactChild;
    /** className - CSS class name for the submenu item */
    className?: string;
    /** isDisabled - whether the submenu item is disabled */
    isDisabled?: boolean;
    /** onClick - function called when the submenu item is clicked */
    onClick?: (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
    /** onClick - function called when a key is pressed within the submenu item */
    onKeyDown?: (event: React.KeyboardEvent<HTMLLIElement>) => void;
    /** onClick - function called when the mouse enters the submenu item */
    onMouseEnter?: (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
    /** onClick - function called when the mouse leaves the submenu item */
    onMouseLeave?: (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
    /** onOpen - function called when the submenu opens */
    onOpen?: () => void;
    /** ref - reference to the menu element in the DOM */
    ref?: string | ((instance: HTMLLIElement | null) => void) | React.RefObject<HTMLLIElement> | null | undefined;
    /** rightBoundaryElement - an HTMLElement defining the right boundary for the submenu item */
    rightBoundaryElement?: HTMLElement;
    /** role - ARIA role for the submenu item */
    role?: string;
    /** tabIndex - indicator of whether the submenu item is focusable */
    tabIndex?: number;
}

interface SubmenuItemState {
    /** isSubmenuOpen - whether the submenu is open */
    isSubmenuOpen: boolean;
    /** isSubmenuOpen - index of the focused submenu item */
    submenuFocusIndex: number | null | undefined;
}

/**
 * A menu-item component which triggers open a submenu
 *
 * @NOTE: Nested submenus are NOT currently supported, switching
 * focus with arrow keys in the subsubmenu is not working properly.
 */
class SubmenuItem extends React.Component<SubmenuItemProps, SubmenuItemState> {
    state = {
        isSubmenuOpen: false,
        submenuFocusIndex: null,
    };

    getMenuAlignmentClasses = (): Record<string, boolean> => {
        if (!this.submenuTriggerEl || !this.submenuEl) {
            return {};
        }

        const { rightBoundaryElement, bottomBoundaryElement } = this.props;
        const submenuElBounding = this.submenuEl.getBoundingClientRect();
        const submenuTriggerElBounding = this.submenuTriggerEl.getBoundingClientRect();
        const rightBoundaryElementBounding = rightBoundaryElement
            ? rightBoundaryElement.getBoundingClientRect()
            : { right: window.innerWidth };
        const bottomBoundaryElementBounding = bottomBoundaryElement
            ? bottomBoundaryElement.getBoundingClientRect()
            : { bottom: window.innerHeight };

        const isLeftAligned =
            submenuTriggerElBounding.right + submenuElBounding.width > rightBoundaryElementBounding.right;
        const isBottomAligned =
            submenuTriggerElBounding.top + submenuElBounding.height > bottomBoundaryElementBounding.bottom;

        return {
            [SUBMENU_LEFT_ALIGNED_CLASS]: isLeftAligned,
            [SUBMENU_BOTTOM_ALIGNED_CLASS]: isBottomAligned,
        };
    };

    handleMenuItemClick = (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
        const { isDisabled, onClick } = this.props;

        // If aria-disabled is passed as a prop, we should ignore clicks on this menu item
        if (isDisabled) {
            event.stopPropagation();
            event.preventDefault();

            return;
        }

        if (onClick) {
            onClick(event);
        }
    };

    submenuTriggerEl: HTMLElement | null | undefined;

    submenuEl: HTMLElement | null | undefined;

    handleKeyDown = (event: React.KeyboardEvent<HTMLLIElement>) => {
        switch (event.key) {
            case ' ':
            case 'Enter':
            case 'ArrowRight':
                event.stopPropagation();
                event.preventDefault();
                this.openSubmenuAndFocus();
                break;
            default:
                break;
        }
    };

    closeSubmenu = debounce(() => {
        this.setState({
            isSubmenuOpen: false,
        });
    }, 50);

    closeSubmenuAndFocusTrigger = (isKeyboardEvent: boolean | null | undefined) => {
        this.closeSubmenu();
        if (this.submenuTriggerEl && isKeyboardEvent) {
            this.submenuTriggerEl.focus();
        }
    };

    openSubmenu = () => {
        this.closeSubmenu.cancel();
        const { onOpen } = this.props;
        if (onOpen) {
            onOpen();
        }
        this.setState({
            isSubmenuOpen: true,
            submenuFocusIndex: null,
        });
    };

    openSubmenuAndFocus = () => {
        const { onOpen } = this.props;
        if (onOpen) {
            onOpen();
        }

        this.setState({
            isSubmenuOpen: true,
            submenuFocusIndex: 0,
        });
    };

    render() {
        const { children, className, isDisabled, ...rest } = this.props;
        const { isSubmenuOpen, submenuFocusIndex } = this.state;

        const elements = React.Children.toArray(children);
        const submenuTriggerContent = elements[0];
        const submenu = elements[1] as React.ReactElement;
        if (elements.length !== 2 || !submenuTriggerContent || !submenu) {
            throw new Error('SubmenuItem must have exactly two children, a trigger component and a <Menu>');
        }

        const chevron = <Arrow16 className="menu-item-arrow" width={12} height={12} />;

        const menuItemProps: SubmenuItemProps = {
            ...omit(rest, ['bottomBoundaryElement', 'onClick', 'onOpen', 'rightBoundaryElement', 'role', 'tabIndex']),
            'aria-disabled': isDisabled ? 'true' : undefined,
            'aria-expanded': isSubmenuOpen ? 'true' : 'false',
            'aria-haspopup': 'true',

            className: classNames('menu-item', 'submenu-target', className),
            onClick: this.handleMenuItemClick,
            onMouseLeave: this.closeSubmenu,
            onMouseEnter: this.openSubmenu,
            onKeyDown: this.handleKeyDown,
            ref: (ref: HTMLLIElement | null) => {
                this.submenuTriggerEl = ref;
            },
            role: 'menuitem',
            tabIndex: -1,
        };

        const submenuProps = {
            className: classNames(submenu.props.className, 'submenu', this.getMenuAlignmentClasses()),
            initialFocusIndex: submenuFocusIndex,
            // Hide the menu instead of unmounting it. Otherwise onMouseLeave won't work.
            isHidden: !isSubmenuOpen,
            isSubmenu: true,
            onClose: this.closeSubmenuAndFocusTrigger,
            setRef: (ref: HTMLLIElement) => {
                this.submenuEl = ref;
            },
        };

        return (
            <li {...menuItemProps}>
                {submenuTriggerContent}
                {chevron}
                {React.cloneElement(submenu, submenuProps)}
            </li>
        );
    }
}

export default SubmenuItem;
