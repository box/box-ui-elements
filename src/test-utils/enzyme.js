/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import { mount as baseMount, shallow as baseShallow } from 'enzyme';

// Global providers that need setup
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../styles/theme';

// Wrap this around every component so they can use app-wide context.
// Prevents breaking due to missing providers.
// See https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/getWrappingComponent.md
/* eslint-disable react/prop-types */
const Wrappers = ({ children }) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

/**
 * mount() from Enzyme but with required app-wide context providers included by default
 */
const mountConnected = (element, options = {}) => baseMount(element, { wrappingComponent: Wrappers, ...options });

/**
 * shallow() from Enzyme but with required app-wide context providers included by default.
 */
const shallowConnected = (element, options = {}) => baseShallow(element, { wrappingComponent: Wrappers, ...options });

// Re-export everything from Enzyme
export * from 'enzyme';
// Export wrapped methods
export { mountConnected, shallowConnected };
