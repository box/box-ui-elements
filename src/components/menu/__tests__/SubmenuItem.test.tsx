import React from 'react';
import { shallow } from 'enzyme';

import SubmenuItem from '../SubmenuItem';

jest.mock('lodash/debounce', () =>
    jest.fn(fn => {
        fn.cancel = jest.fn();
        return fn;
    }),
);

describe('components/menu/SubmenuItem', () => {
    const getWrapper = (props = {}) => {
        return shallow<SubmenuItem>(
            <SubmenuItem {...props}>
                <div />
                <div />
            </SubmenuItem>,
        );
    };

    describe('render()', () => {
        test('should correctly render default component', () => {
            const wrapper = getWrapper();

            expect(wrapper).toMatchSnapshot();
        });

        test('should correctly render correctly when isDisabled is true', () => {
            const wrapper = getWrapper({
                isDisabled: true,
            });

            expect(wrapper).toMatchSnapshot();
        });

        test('should correctly render correctly when isSubmenuOpen is true', () => {
            const wrapper = getWrapper({});
            wrapper.setState({
                isSubmenuOpen: true,
            });

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('openSubmenu()', () => {
        test('should set isSubmenuOpen state to true', () => {
            const wrapper = getWrapper();
            wrapper.instance().openSubmenu();

            expect(wrapper.state('isSubmenuOpen')).toBe(true);
        });
    });

    describe('closeSubmenuAndFocusTrigger()', () => {
        test('should set isSubmenuOpen state to false and not focus trigger element if not keyboard event', () => {
            const wrapper = getWrapper();
            const instance = wrapper.instance();
            const focusSpy = jest.fn();
            wrapper.setState({
                isSubmenuOpen: true,
            });
            instance.submenuTriggerEl = document.createElement('input');
            instance.submenuTriggerEl.focus = focusSpy;

            instance.closeSubmenuAndFocusTrigger(false);

            expect(wrapper.state('isSubmenuOpen')).toBe(false);
            expect(focusSpy).not.toHaveBeenCalled();
        });

        test('should set isSubmenuOpen state to false and focus trigger element if keyboard event', () => {
            const wrapper = getWrapper();
            const instance = wrapper.instance();
            const focusSpy = jest.fn();
            wrapper.setState({
                isSubmenuOpen: true,
            });
            instance.submenuTriggerEl = document.createElement('input');
            instance.submenuTriggerEl.focus = focusSpy;

            instance.closeSubmenuAndFocusTrigger(true);

            expect(wrapper.state('isSubmenuOpen')).toBe(false);
            expect(focusSpy).toHaveBeenCalled();
        });
    });

    describe('handleKeyDown()', () => {
        test.each([' ', 'Enter', 'ArrowRight'])('should call openSubmenu() when the %s key is pressed', key => {
            const wrapper = getWrapper();
            const instance = wrapper.instance();
            const openSubmenuSpy = jest.fn();
            const stopPropagationSpy = jest.fn();
            const preventDefaultSpy = jest.fn();
            instance.openSubmenuAndFocus = openSubmenuSpy;
            wrapper.find('li').simulate('keydown', {
                key,
                stopPropagation: stopPropagationSpy,
                preventDefault: preventDefaultSpy,
            });
            expect(openSubmenuSpy).toHaveBeenCalled();
            expect(stopPropagationSpy).toHaveBeenCalled();
            expect(preventDefaultSpy).toHaveBeenCalled();
        });
    });

    describe('handleMenuItemClick()', () => {
        test('should call onClick() when isDisable is false', () => {
            const onClickSpy = jest.fn();
            const wrapper = getWrapper({
                isDisabled: false,
                onClick: onClickSpy,
            });
            wrapper.find('li').simulate('click', {});
            expect(onClickSpy).toHaveBeenCalled();
        });

        test('should not call onClick() and stop propagation and prevent default when isDisable is true', () => {
            const onClickSpy = jest.fn();
            const stopPropagationSpy = jest.fn();
            const preventDefaultSpy = jest.fn();
            const wrapper = getWrapper({
                isDisabled: true,
                onClick: onClickSpy,
            });
            wrapper.find('li').simulate('click', {
                stopPropagation: stopPropagationSpy,
                preventDefault: preventDefaultSpy,
            });
            expect(onClickSpy).not.toHaveBeenCalled();
            expect(stopPropagationSpy).toHaveBeenCalled();
            expect(preventDefaultSpy).toHaveBeenCalled();
        });
    });

    describe('getMenuAlignmentClasses()', () => {
        const SUBMENU_LEFT_ALIGNED_CLASS = 'is-left-aligned';
        const SUBMENU_BOTTOM_ALIGNED_CLASS = 'is-bottom-aligned';

        test.each`
            alignment   | specifiedElement           | submenuElRect      | submenuTriggerElRect                     | expectedClass
            ${'left'}   | ${'rightBoundaryElement'}  | ${{ width: 600 }}  | ${{ right: 100, bottom: 200 }}           | ${SUBMENU_LEFT_ALIGNED_CLASS}
            ${'bottom'} | ${'bottomBoundaryElement'} | ${{ height: 600 }} | ${{ right: 100, bottom: 200, top: 100 }} | ${SUBMENU_BOTTOM_ALIGNED_CLASS}
        `(
            'should set $expectedClass to true when submenu should be $alignment aligned and $specifiedElement is specified',
            ({ specifiedElement, submenuElRect, submenuTriggerElRect, expectedClass }) => {
                const wrapper = getWrapper({
                    [specifiedElement]: {
                        getBoundingClientRect: jest.fn(() => ({
                            right: 500,
                            bottom: 500,
                        })),
                    },
                });
                const instance = wrapper.instance();
                instance.submenuEl = document.createElement('li');
                instance.submenuEl.getBoundingClientRect = jest.fn(() => submenuElRect as DOMRect);
                instance.submenuTriggerEl = document.createElement('li');
                instance.submenuTriggerEl.getBoundingClientRect = jest.fn(() => submenuTriggerElRect as DOMRect);
                const classes = instance.getMenuAlignmentClasses();
                expect(classes[expectedClass]).toBe(true);
            },
        );

        test.each`
            alignment   | unspecifiedElement         | submenuElRect      | submenuTriggerElRect                     | expectedClass
            ${'left'}   | ${'rightBoundaryElement'}  | ${{ width: 600 }}  | ${{ right: 100, bottom: 200 }}           | ${SUBMENU_LEFT_ALIGNED_CLASS}
            ${'bottom'} | ${'bottomBoundaryElement'} | ${{ height: 600 }} | ${{ right: 100, bottom: 200, top: 100 }} | ${SUBMENU_BOTTOM_ALIGNED_CLASS}
        `(
            'should set SUBMENU_LEFT_ALIGNED_CLASS to true when submenu should be $alignment aligned and $unspecifiedElement is not specified',
            ({ submenuElRect, submenuTriggerElRect, expectedClass }) => {
                const wrapper = getWrapper();
                Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 500 });
                Object.defineProperty(window, 'innerHeight', { writable: true, configurable: true, value: 500 });
                const instance = wrapper.instance();
                instance.submenuEl = document.createElement('li');
                instance.submenuEl.getBoundingClientRect = jest.fn(() => submenuElRect as DOMRect);
                instance.submenuTriggerEl = document.createElement('li');
                instance.submenuTriggerEl.getBoundingClientRect = jest.fn(() => submenuTriggerElRect as DOMRect);
                const classes = instance.getMenuAlignmentClasses();
                expect(classes[expectedClass]).toBe(true);
            },
        );
    });
});
