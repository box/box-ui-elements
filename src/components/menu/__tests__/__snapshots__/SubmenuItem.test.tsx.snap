// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/menu/SubmenuItem render() should correctly render correctly when isDisabled is true 1`] = `
<li
  aria-disabled="true"
  aria-expanded="false"
  aria-haspopup="true"
  className="menu-item submenu-target"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  <div
    key=".0"
  />
  <Arrow16
    className="menu-item-arrow"
    height={12}
    width={12}
  />
  <div
    className="submenu"
    initialFocusIndex={null}
    isHidden={true}
    isSubmenu={true}
    key=".1"
    onClose={[Function]}
    setRef={[Function]}
  />
</li>
`;

exports[`components/menu/SubmenuItem render() should correctly render correctly when isSubmenuOpen is true 1`] = `
<li
  aria-expanded="true"
  aria-haspopup="true"
  className="menu-item submenu-target"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  <div
    key=".0"
  />
  <Arrow16
    className="menu-item-arrow"
    height={12}
    width={12}
  />
  <div
    className="submenu"
    initialFocusIndex={null}
    isHidden={false}
    isSubmenu={true}
    key=".1"
    onClose={[Function]}
    setRef={[Function]}
  />
</li>
`;

exports[`components/menu/SubmenuItem render() should correctly render default component 1`] = `
<li
  aria-expanded="false"
  aria-haspopup="true"
  className="menu-item submenu-target"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  <div
    key=".0"
  />
  <Arrow16
    className="menu-item-arrow"
    height={12}
    width={12}
  />
  <div
    className="submenu"
    initialFocusIndex={null}
    isHidden={true}
    isSubmenu={true}
    key=".1"
    onClose={[Function]}
    setRef={[Function]}
  />
</li>
`;
